### PR TITLE
Bigtable: instrument client with Opencensus metrics

### PR DESCRIPTION
**Pull request by igorbernstein2**
It adds stats to logical operations for things like:
- how long did logically method call take across all of its retry attempts
- whats the distribution of rows read per operation
- whats the distribution of the number of entries in a bulk mutation

This should allow users to better debug the performance of the client. Allowing to debug scenarios like high perceived latency on the client side, but no latency visible on the serverside due to retries.

This also adds a more user friendly way to enable data collection by exposing a method on the settings class.

Main changes:
* Follow grpc's example and create Rpc*Constant files to configure opencensus stat views
* Mark the constants as internal and expose them via a static method on the settings
* Create callables for unary chains and special callables for ReadRows & MutateRows
* Push metric & tracing instrumentation to the leaf callables
*

### CODE REVIEW

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 1]
```diff
@@ -208,17 +244,35 @@ public static EnhancedBigtableStub create(EnhancedBigtableStubSettings settings)
...
+                    .setIdleTimeout(settings.readRowSettings().getRetrySettings().getTotalTimeout())
+                    .build(),
+                rowAdapter)
+            .first();
+
+    UnaryCallable<Query, RowT> traced =
+        new TracedUnaryCallable<>(
+            readRowCallable,
+            clientContext.getTracerFactory(),
+            SpanName.of(TRACING_OUTER_CLIENT_NAME, "ReadRows"));
```
Comment (rahulKQL) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: I think here & below SpanName name's suffix should be `"ReadRow"`

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 2]
```diff
@@ -208,17 +244,35 @@ public static EnhancedBigtableStub create(EnhancedBigtableStubSettings settings)
...
+
+    UnaryCallable<Query, RowT> measured =
+        new MeasuredUnaryCallable<>(
+            traced,
+            TRACING_OUTER_CLIENT_NAME + ".ReadRows",
+            tagger,
+            statsRecorder,
+            clientContext.getClock());
+
+    return measured.withDefaultCallContext(clientContext.getDefaultCallContext());
```
Comment (rahulKQL) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Can we use `return createUserFacingUnaryCallable("ReadRow", readRowCallable);` ?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 3]
```diff
@@ -426,24 +519,11 @@ public static EnhancedBigtableStub create(EnhancedBigtableStubSettings settings)
...
-          String methodName, ServerStreamingCallable<RequestT, ResponseT> inner) {
-
-    ServerStreamingCallable<RequestT, ResponseT> traced =
-        new TracedServerStreamingCallable<>(
-            inner,
-            clientContext.getTracerFactory(),
-            SpanName.of(TRACING_OUTER_CLIENT_NAME, methodName));
+    UnaryCallable<RequestT, ResponseT> measured =
+        new MeasuredUnaryCallable<>(
+            traced, methodName, tagger, statsRecorder, clientContext.getClock());
```
Comment (rahulKQL) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Should the methodName parameter include `TRACING_OUTER_CLIENT_NAME`? considering this as user-facing.

File: google-cloud-clients/google-cloud-bigtable/README.md [hunk 4]
```diff
@@ -223,6 +224,83 @@ Tracing.getTraceConfig().updateActiveTraceParams(
...
+  operation.
+
+* `cloud.google.com/java/bigtable/rows_per_op`: A distribution of rows
+  read per ReadRows operation across all retry attempts.
+
+* `cloud.google.com/java/bigtable/mutations_per_batch`: A distribution
+  of mutations per BulkMutation.
+
+
+By default, the functionality is disabled. To enable, you need to add a couple of
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/README.md]: I would reword: "To enable, please follow the instructions below:"

File: google-cloud-clients/google-cloud-bigtable/README.md [hunk 5]
```diff
@@ -223,6 +224,83 @@ Tracing.getTraceConfig().updateActiveTraceParams(
...
+compile 'io.opencensus:opencensus-impl:0.23.0'
+compile 'io.opencensus:opencensus-exporter-stats-stackdriver:0.23.0'
+```
+If you are using SBT, add this to your dependencies
+```Scala
+libraryDependencies += "io.opencensus" % "opencensus-impl" % "0.23.0"
+libraryDependencies += "io.opencensus" % "opencensus-exporter-stats-stackdriver" % "0.23.0"
+```
+
+Then at the start of your application configure the exporter and enable the Bigtable stats views:
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/README.md]: Remove "then"

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java [hunk 6]
```diff
@@ -124,6 +125,48 @@ public ManagedChannelBuilder apply(ManagedChannelBuilder input) {
     return builder;
   }
 
+  /**
+   * Enables Opencensus metric aggregations.
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java]: minor - change all occurrences of "Opencensus" to "OpenCensus"

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java [hunk 7]
```diff
@@ -124,6 +125,48 @@ public ManagedChannelBuilder apply(ManagedChannelBuilder input) {
...
+   * </pre>
+   *
+   * Java:
+   * <pre>{@code
+   *   StackdriverStatsExporter.createAndRegister();
+   *   BigtableDataSettings.enableOpencensusStats();
+   * }</pre>
+   */
+  @BetaApi("Opencensus stats integration is currently unstable and may change in the future")
+  public static void enableOpencensusStats() {
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java]: OpenCensus

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 8]
```diff
@@ -340,8 +401,10 @@ public static EnhancedBigtableStub create(EnhancedBigtableStubSettings settings)
    *   <li>Wrap batch failures in a {@link
    *       com.google.cloud.bigtable.data.v2.models.MutateRowsException}.
    *   <li>Split the responses using {@link MutateRowsBatchingDescriptor}.
+   *   <li>Add tracing & metrics.
    * </ul>
    */
+  @Deprecated
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: point to new method?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/MeasuredMutateRowsCallable.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import io.opencensus.stats.StatsRecorder;
+import io.opencensus.tags.TagContext;
+import io.opencensus.tags.TagValue;
+import io.opencensus.tags.Tagger;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+
+/**
+ * @deprecated Please use {@link MeasuredMutateRowsCallableV2}.
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/MeasuredMutateRowsCallable.java]: please add an internal use only comment as well

Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/MeasuredMutateRowsCallable.java]: done

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 8 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: done

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java (hunk 7 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java]: done

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java (hunk 6 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataSettings.java]: done

File: google-cloud-clients/google-cloud-bigtable/README.md (hunk 5 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/README.md]: done

File: google-cloud-clients/google-cloud-bigtable/README.md (hunk 4 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/README.md]: changed the wording, PTAL
